tensorly/backend/core.py

```diff
@@ -644,56 +640,20 @@ def kron(self, a, b):
         return self.reshape(a * b, (s1 * s3, s2 * s4))
 
     def kr(self, matrices, weights=None, mask=None):
-        """Khatri-Rao product of a list of matrices
+        n_columns = matrices[0].shape[1]
+        n_factors = len(matrices)
 
-        This can be seen as a column-wise kronecker product.
+        start = ord('a')
+        common_dim = 'z'
+        target = ''.join(chr(start + i) for i in range(n_factors))
+        source = ','.join(i + common_dim for i in target)
+        operation = source + '->' + target + common_dim
 
-        Parameters
-        ----------
-        matrices : list of tensors
-            List of 2D tensors with the same number of columns, i.e.::
+        if weights is not None:
+            matrices = [m if i else m*self.reshape(weights, (1, -1)) for i, m in enumerate(matrices)]
 
-                for i in len(matrices):
-                    matrices[i].shape = (n_i, m)
-
-        Returns
-        -------
-        khatri_rao_product : tensor of shape ``(prod(n_i), m)``
-            Where ``prod(n_i) = prod([m.shape[0] for m in matrices])`` (i.e. the
-            product of the number of rows of all the matrices in the product.)
-
-        Notes
-        -----
-        Mathematically:
-
-        .. math::
-            \\text{If every matrix } U_k \\text{ is of size } (I_k \\times R),\\\\
-            \\text{Then } \\left(U_1 \\bigodot \\cdots \\bigodot U_n \\right) \\\\
-            text{ is of size } (\\prod_{k=1}^n I_k \\times R)
-        """
-        if len(matrices) < 2:
-            raise ValueError('kr requires a list of at least 2 matrices, but {} '
-                            'given.'.format(len(matrices)))
-
-        n_col = self.shape(matrices[0])[1]
-        for i, e in enumerate(matrices[1:]):
-            if not i:
-                if weights is None:
-                    res = matrices[0]
-                else:
-                    res = matrices[0]*self.reshape(weights, (1, -1))
-            s1, s2 = self.shape(res)
-            s3, s4 = self.shape(e)
-            if not s2 == s4 == n_col:
-                raise ValueError('All matrices should have the same number of columns.')
-
-            a = self.reshape(res, (s1, 1, s2))
-            b = self.reshape(e, (1, s3, s4))
-            res = self.reshape(a * b, (-1, n_col))
-
-        m = self.reshape(mask, (-1, 1)) if mask is not None else 1
-
-        return res*m
+        m = mask.reshape((-1, 1)) if mask is not None else 1
```

Comment:
tl.kr wasn’t NumPy specific:?, rewriting it with einsum will make it slower in some backends.

Reply:
My thinking was the `mxnet.np` interface will provide SVD and einsum methods, and so `kr` might not need to be backend-defined at all after that change. We've had a backend-specific bug before with the masking feature and so, absent other priorities, I figure it's better to have everything as unified as possible.

You're right that the einsum performance is likely very different for each backend. I'll revert the `kr` changes. It might be interesting to benchmark the two implemenetations if/when we can change `mxnet` over to the new interface.

Reply:
Thanks for the input!

Reply:
Indeed, whether we implement it using einsum or the current version, it will still have to be defined in the backend unless each backend implements it. That implementation was also much faster when I wrote it than other alternatives.

In any case, I also wrote an [`einsum` version](https://github.com/tensorly/tensorly/blob/master/tensorly/tenalg/einsum_tenalg/_khatri_rao.py), which is automatically used if you set the tenalg backend to "einsum" (`tensorly.tenalg.set_tenalg_backend("einsum")`).

Reply:
If kr's implementation is set in the frontend, perhaps we can just remove T.kr from the backend altogether?